client/src/components/Clan/RoleAssignmentModal.tsx

```diff
@@ -0,0 +1,147 @@
+import { useTranslation } from 'react-i18next';
+import FantasyButton from '../FantasyButton';
+import Text from '../Text';
+import { useAlert } from '../../hooks/useAlert';
+import Server from '../../utils/Server';
+import catchError from '../../utils/catchError';
+
+interface Member {
+  id: string;
+  name: string;
+}
+
+interface Role {
+  id: string;
+  name: string;
+}
+
+interface RoleAssignmentModalProps {
+  open: boolean;
+  onClose: () => void;
+  clanId: string;
+  role: Role | null;
+  members: Member[];
+  onAssigned: () => void;
+}
+
+const RoleAssignmentModal: React.FC<RoleAssignmentModalProps> = ({
+  open,
+  onClose,
+  clanId,
+  role,
+  members,
+  onAssigned,
+}) => {
+  const { t } = useTranslation();
+  const Alert = useAlert();
+
+  const [selectedMembers, setSelectedMembers] = useState<string[]>([]);
+
+  useEffect(() => {
+    if (open) {
+      setSelectedMembers([]);
+    }
+  }, [open, members]);
+
+  const handleAssign = () => {
+    if (selectedMembers.length === 0 || !role) {
+      Alert.open('error', t('pleaseSelectMember'));
+      return;
+    }
+
+    // Assign role to all selected members
+    const promises = selectedMembers.map(
+      (memberId) => Server.Clan.assignRole(clanId, memberId, role.id),
+    );
+
+    Promise.all(promises)
+      .then(() => {
+        Alert.open('success', t('rolesAssigned', { count: selectedMembers.length }));
+        setSelectedMembers([]);
+        onAssigned();
+        onClose();
+      })
+      .catch(catchError(Alert));
+  };
+
+  if (!open || !role) return null;
+
+  return (
+    <Box
+      sx={{
+        position: 'fixed',
+        top: 0,
+        left: 0,
+        right: 0,
+        bottom: 0,
+        bgcolor: 'rgba(0, 0, 0, 0.8)',
+        display: 'flex',
+        alignItems: 'center',
+        justifyContent: 'center',
+        zIndex: 1400,
+      }}
+      onClick={onClose}
+    >
+      <Paper
+        onClick={(e) => e.stopPropagation()}
+        sx={{
+          maxWidth: 500,
+          width: '90%',
+          p: 3,
+        }}
+      >
+        <Text h3 bold color="secondary" sx={{ mb: 2 }}>
+          {t('assignRole')}: {role.name}
+        </Text>
+
+        {members.length === 0 ? (
+          <Text color="text.secondary" sx={{ mb: 3 }}>
+            {t('noMembersAvailable')}
+          </Text>
+        ) : (
+          <FormControl fullWidth sx={{ mb: 3 }}>
+            <InputLabel>{t('selectMembers')}</InputLabel>
+            <Select
+              multiple
+              value={selectedMembers}
+              onChange={(e) => setSelectedMembers(e.target.value as string[])}
+              label={t('selectMembers')}
+              renderValue={(selected) => {
+                const selectedNames = members
+                  .filter((m) => selected.includes(m.id))
+                  .map((m) => m.name)
+                  .join(', ');
+                return selectedNames || t('selectMembers');
+              }}
```

Comment:
I prefer having handlers outside of the render function, so it would be better to move the handlers above.